### PR TITLE
[spring-boot-hw-bookstore] added spring boot security

### PR DESCRIPTION
**Pull request by Andriy-Hrynevetskyi**
list 0f changes:
- spring security starter
-SecurityConfig
-implemented role access
-added role entitty and all neccessary services dto repositories...
- implemented CustomUserDetailsService 
- added liquibase scripts

### CODE REVIEW

**Review by HovorukhaBohdan — APPROVED** [review]
Good job!

File: src/main/java/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth.requestMatchers("/auth/**","/error","/swagger-ui/**")
```
Comment (d1sam) [inline src/main/java/bookstore/config/SecurityConfig.java]:
```suggestion
                        auth -> auth.requestMatchers("/auth/**","/error","/swagger-ui/**", "/v3/api-docs/**")
```

File: src/main/java/bookstore/model/User.java [hunk 2]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java]:
```suggestion
        return !deleted;
```

File: src/main/java/bookstore/model/User.java [hunk 3]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java]:
```suggestion
        return !deleted;
```

File: src/main/java/bookstore/model/User.java [hunk 4]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java]:
```suggestion
        return !deleted;
```

File: src/main/java/bookstore/model/User.java [hunk 5]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java]:
```suggestion
        return !deleted;
```

File: src/main/java/bookstore/model/User.java [hunk 6]
```diff
@@ -24,4 +32,44 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles = new HashSet<>();
+
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java:46]:
```suggestion
  @Column(nullable = false)
  private boolean deleted = false;
```

File: src/main/java/bookstore/model/User.java [hunk 7]
```diff
@@ -5,13 +5,21 @@
...
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 import lombok.Data;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Entity
 @Table(name = "users")
 @Data
```
Comment (d1sam) [inline src/main/java/bookstore/model/User.java:25]:
```suggestion
@Data
@SQLDelete(sql = "UPDATE users SET deleted = true WHERE id=?")
@Where(clause = "deleted=false")
```

File: src/main/java/bookstore/repository/role/RoleRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package bookstore.repository.role;
+
+import bookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (d1sam) [inline src/main/java/bookstore/repository/role/RoleRepository.java:8]:
```suggestion
@Repository
public interface RoleRepository extends JpaRepository<Role, Long> {
```

File: src/main/java/bookstore/service/user/UserServiceImpl.java [hunk 9]
```diff
@@ -7,20 +7,27 @@
...
         if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
             throw new RegistrationException("User wit given email already exists");
         }
-        User user = userMapper.toModel(requestDto);
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(passwordEncoder.encode(requestDto.getRepeatPassword()));
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (d1sam) [inline src/main/java/bookstore/service/user/UserServiceImpl.java]: use mapper

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+                    foreignKeyName: users_id
+                    nullable: false
+                    references: users(id)
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: roles_id
+                    nullable: false
+                    references: roles(id)
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: add empty line to the end of file

File: src/main/resources/db/changelog/changes/05-add-role-id-column-to-users-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/05-add-role-id-column-to-users-table.yaml:1]: why do you need this column? you have `users_roles` table

File: src/main/resources/db/changelog/changes/06-insert-users.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/06-insert-users.yaml:1]: you should add roles to users by users_roles table and use valueComputed

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done according to my comment

File: src/main/resources/db/changelog/changes/07-insert-values-in-users-roles-table.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+                  valueComputed: (SELECT id FROM roles WHERE role='ADMIN')
+        - insert:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  valueComputed: (SELECT id FROM users WHERE email='user@mail.com')
+              - column:
+                  name: role_id
+                  valueComputed: (SELECT id FROM roles WHERE role='USER')
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/07-insert-values-in-users-roles-table.yaml]: please add empty line at the end of file

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml:1]: `users_roles` table should have composite key on columns: `user_id` and `role_id`.

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but please add composite key to `users_roles` table on columns `user_id` and `role_id`

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml
Comment (d1sam) [reply src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml:1]: well done, but please pay attention to the empty line at the end of file :)

**Review by d1sam — APPROVED** [review]

**Pull request merged**